flycheck.el

```diff
@@ -2742,54 +2788,105 @@ current syntax check."
   (flycheck-error-list-refresh)
   (flycheck-hide-error-buffer))
 
-(defun flycheck-teardown ()
-  "Teardown Flycheck in the current buffer..
+(defun flycheck--empty-variables ()
+  "Empty variables used by Flycheck."
+  (kill-local-variable 'flycheck--idle-trigger-timer)
+  (kill-local-variable 'flycheck--idle-trigger-conditions))
+
+(defun flycheck-teardown (&optional ignore-global)
+  "Teardown Flycheck in the current buffer.
 
 Completely clear the whole Flycheck state.  Remove overlays, kill
-running checks, and empty all variables used by Flycheck."
+running checks, and empty all variables used by Flycheck.
+
+Unless optional argument IGNORE-GLOBAL is non-nil, check to see
+if no more Flycheck buffers remain (aside from the current
+buffer), and if so then clean up global hooks."
   (flycheck-safe-delete-temporaries)
   (flycheck-stop)
   (flycheck-clean-deferred-check)
   (flycheck-clear)
-  (flycheck-cancel-error-display-error-at-point-timer))
+  (flycheck-cancel-error-display-error-at-point-timer)
+  (flycheck--clear-idle-trigger-timer)
+  (flycheck--empty-variables)
+  (unless (or ignore-global
+              (->> (buffer-list)
+                   (seq-remove (apply-partially #'equal (current-buffer)))
+                   (seq-some (apply-partially #'buffer-local-value
+                                              'flycheck-mode))))
```

Comment:
This looks a bit convoluted, and it traverses the buffer list twice. Would this do the same?

```
(seq-some (lambda (buf) (and (not (equal buf (current-buffer)))
                        (buffer-local-value 'flycheck-mode buf)))
          (buffer-list))
```